psl-core/src/main/java/org/linqs/psl/application/inference/online/OnlineServer.java

```diff
@@ -0,0 +1,286 @@
+ * limitations under the License.
+ */
+package org.linqs.psl.application.inference.online;
+
+import org.linqs.psl.application.inference.online.messages.OnlineMessage;
+import org.linqs.psl.application.inference.online.messages.actions.controls.Exit;
+import org.linqs.psl.application.inference.online.messages.actions.controls.Stop;
+import org.linqs.psl.application.inference.online.messages.responses.ActionStatus;
+import org.linqs.psl.application.inference.online.messages.responses.OnlineResponse;
+import org.linqs.psl.config.Options;
+import org.linqs.psl.model.rule.Rule;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.EOFException;
+import java.io.File;
+import java.io.IOException;
+import java.io.ObjectInputStream;
+import java.io.ObjectOutputStream;
+import java.net.ServerSocket;
+import java.net.Socket;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+import java.util.UUID;
+import java.util.concurrent.BlockingQueue;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
+import java.util.concurrent.LinkedBlockingQueue;
+
+/**
+ * A class for listening for new client connections, queueing actions from OnlineClients, and sending OnlineResponses.
+ */
+public class OnlineServer {
+    private static final Logger log = LoggerFactory.getLogger(OnlineServer.class);
+
+    private boolean listening;
+    private ServerConnectionThread serverThread;
+    private Set<ClientConnectionThread> clientConnectionThreads;
+    private BlockingQueue<OnlineMessage> queue;
+    private ConcurrentMap<UUID, ClientConnectionThread> messageIDConnectionMap;
+
+    public OnlineServer() {
+        listening = false;
+        serverThread = new ServerConnectionThread();
+        queue = new LinkedBlockingQueue<OnlineMessage>();
+        messageIDConnectionMap = new ConcurrentHashMap<UUID, ClientConnectionThread>();
+        clientConnectionThreads = new HashSet<ClientConnectionThread>();
+    }
+
+    /**
+     * Start up the server and listen for new connections on the configured port.
+     * This method does not block. A new thread is started to wait for connections.
+     */
+    public void start() {
+        listening = true;
+        serverThread.start();
+    }
+
+    /**
+     * Get the next action from the client.
+     * This method will block until an action is available to take from the queue.
+     */
+    public OnlineMessage getAction() {
+        OnlineMessage nextAction = null;
+
+        while (nextAction == null) {
+            try {
+                nextAction = queue.take();
+            } catch (InterruptedException ex) {
+                log.warn("Interrupted while taking an online action from the queue.", ex);
+                return null;
+            }
+
+            if (nextAction instanceof Exit) {
+                onActionExecution(nextAction, new ActionStatus(nextAction, true, "Session Closed."));
+                nextAction = null;
```

Comment:
What's the reason here for not breaking?
To deplete the queue?
Multiple clients?
Race conditions?

We'll probably need to chat in-person about it.

Reply:
The reason for not breaking is there can be multiple client sessions and more actions in the queue. Exit just ends the sessions with a single client. We still want to keep the server alive and be able to have a new client connect and send actions.